client/lib/plans/index.js

```diff
@@ -16,27 +16,34 @@ import {
 	FEATURES_LIST,
 	PLANS_LIST,
 	PLAN_FREE,
-	PLAN_JETPACK_BUSINESS,
-	PLAN_JETPACK_BUSINESS_MONTHLY,
 	PLAN_JETPACK_FREE,
-	PLAN_JETPACK_PERSONAL,
-	PLAN_JETPACK_PERSONAL_MONTHLY,
-	PLAN_JETPACK_PREMIUM,
-	PLAN_JETPACK_PREMIUM_MONTHLY,
 	PLAN_PERSONAL,
 } from 'lib/plans/constants';
+import {
+	TERM_ANNUALLY,
+	TERM_MONTHLY,
+	TYPE_BUSINESS,
+	TYPE_FREE,
+	TYPE_PERSONAL,
+	TYPE_PREMIUM,
+} from './constants';
 
 /**
  * Module vars
  */
 const isPersonalPlanEnabled = isEnabled( 'plans/personal-plan' );
 
-export function isFreePlan( plan ) {
-	return plan === PLAN_FREE || plan === PLAN_JETPACK_FREE;
+export function getPlans() {
+	return PLANS_LIST;
 }
 
-export function getPlan( plan ) {
-	return PLANS_LIST[ plan ];
+export function getPlan( planKey ) {
+	if ( Object.prototype.toString.apply( planKey ) === '[object Object]' ) {
```

Comment:
How about using a lodash function `isPlainObject`? Though, checking if `planKey` is string here would be simpler.

Reply:
I only wanted to make a lookup in PLANS_LIST values for objects and not for e.g. arrays; I like the idea of using `isPlainObject` though - I'll do it in one of follow-up PRs

Reply:
Why do we have to allow for object arguments here anyway? This is the kind of permissive interface and implementation that I think has the potential to get us in trouble. I'd go as far as to say it's a bad enough antipattern that it's worth getting rid of in our legacy code, but certainly something we shouldn't introduce when writing new code.

There might be 'canonical' literature on this, but my main line of argument is that this sort of permissiveness/indecisiveness makes code harder to read (we can't ever be sure what kind of data we're dealing with), and consequentially will lead to duplication of logic/more conditionals -- and in the places where we miss to add those additional conditionals because we overlook them, things are more likely to break.

Shouldn't the consumer of this function know whether we're dealing with a plan slug/key (string) or an object and act accordingly?

Reply:
@ockham we could get rid of it. I introduced it before starting a major calypso-wide refactoring since it looked like it could save a significant amount of time as it looked like a lot of places in the codebase wasn't sure about the argument type. In the end, it turned out that we don't rely on it that much (or at all).

As a side note, I don't think it's that bad though:

> we can't ever be sure what kind of data we're dealing with

Either a plan key (string) or a plan (object), the check seems pretty clear.

> and in the places where we miss to add those additional conditionals because we overlook them, things are more likely to break.

We don't need to add any conditionals out of this function, just use the plan object returned by this function. The return type is always the same (unless it's null).

Anyway - feel free to start a PR and I'll be happy to review :-)